### ISSUE TITLE
Using OSNet for Pedestrian Attribute Recognition

### ISSUE THREAD

**Issue by akashmanna** [KaiyangZhou/deep-person-reid#287]
Hi, I was trying to replicate OSNet for Pedestrian Attribute Recognition on PA-100K dataset as mentioned in the paper. I added a Classifier on top of the base OSNet (2 FC layers, input dims=512 and output dims=number of classes in PA100k), and trained (from scratch) with the values mentioned in the paper, but I am getting F1 Score of around only 50.
Is there any way I can improvise? 
Also, It'll be really helpful if the code used for training on PA-100K is released.

**Comment by KaiyangZhou** [comment]
the model arch is simple, it looks like (using `nn.AdaptiveMaxPool2d(1)`)
```
        # convolutional backbone
        self.conv1 = ConvLayer(3, channels[0], 7, stride=2, padding=3)
        self.maxpool = nn.MaxPool2d(3, stride=2, padding=1)
        self.conv2 = self._make_layer(blocks[0], layers[0], channels[0], channels[1], reduce_spatial_size=True, IN=IN)
        self.conv3 = self._make_layer(blocks[1], layers[1], channels[1], channels[2], reduce_spatial_size=True)
        self.conv4 = self._make_layer(blocks[2], layers[2], channels[2], channels[3], reduce_spatial_size=False)
        self.conv5 = Conv1x1(channels[3], channels[3])
        if pool == 'avg':
            self.global_pool = nn.AdaptiveAvgPool2d(1)
        else:
            self.global_pool = nn.AdaptiveMaxPool2d(1)
        # fully connected layer
        self.fc = self._construct_fc_layer(feature_dim, channels[3], dropout_p=None)
        # classification layer
        self.classifier = nn.Linear(self.feature_dim, num_classes)


def osnet_maxpool(num_classes=1000, loss='softmax', **kwargs):
    return OSNet(num_classes, blocks=[OSBlock, OSBlock, OSBlock], layers=[2, 2, 2],
                 channels=[64, 256, 384, 512], loss=loss, pool='max', **kwargs)
```

I guess the low accuracy problem might be caused by your training code rather than the model code.

I have been very busy so can't find a time to tidy the attribute code. Will try.

**Comment by akashmanna** [comment]
Thanks for the reply. 
The code in this repo used Adaptive Avg Pooling. Earlier we were using the same with a width of 0.5, but following this comment we used Adaptive Max Pool and width=1 and we are getting f1 score of around 0.82 on PA100K dataset.

Looking forward to see your implementation!

**Comment by KaiyangZhou** [comment]
cool

will try to add more features (including attribute recognition) in Mar

**Comment by vakkov** [comment]
@KaiyangZhou Hi! Can you, please, shed some light  into the modifications that need to be made on the training code? Even some dirty but working "draft code" pasted here would be extremely useful. Thanks so much for the support and time you invest in this repo, btw

**Comment by KaiyangZhou** [comment]
@vakkov I'll try to release the attribute code these two weeks!

**Comment by KaiyangZhou** [comment]
@vakkov check `projects/attribute_recognition`

I've not tested the code yet ... let me know if there is any bug

**Comment by vakkov** [comment]
@KaiyangZhou 
Thanks mate, I just noticed the tag; will test it over the weekend and report back!

**Comment by KaiyangZhou** [comment]
@vakkov cool, let me know if there is any bug, feel free to do a pull request if there is

**Comment by vakkov** [comment]
@KaiyangZhou 
those are the evaluation results: `=> Loaded PA100K
  ------------------------------
  subset   | # images
  ------------------------------
  train    |    80000
  val      |    10000
  test     |    10000
  ------------------------------
  total    |   100000
  ------------------------------
  # attributes: 26
  attributes:
      0: Female
      1: AgeOver60
      2: Age18-60
      3: AgeLess18
      4: Front
      5: Side
      6: Back
      7: Hat
      8: Glasses
      9: HandBag
     10: ShoulderBag
     11: Backpack
     12: HoldObjectsInFront
     13: ShortSleeve
     14: LongSleeve
     15: UpperStride
     16: UpperLogo
     17: UpperPlaid
     18: UpperSplice
     19: LowerStripe
     20: LowerPattern
     21: LongCoat
     22: Trousers
     23: Shorts
     24: Skirt&Dress
     25: boots
  ------------------------------
Use plain binary cross entropy
Building model: osnet_maxpool
Model complexity: params=2,206,954 flops=978,891,690
Successfully loaded pretrained weights from "log/pa100k-osnet_maxpool/model-best.pth.tar"
Testing ...
Processed batch 20/313
Processed batch 40/313
Processed batch 60/313
Processed batch 80/313
Processed batch 100/313
Processed batch 120/313
Processed batch 140/313
Processed batch 160/313
Processed batch 180/313
Processed batch 200/313
Processed batch 220/313
Processed batch 240/313
Processed batch 260/313
Processed batch 280/313
Processed batch 300/313
=> BatchTime(s)/BatchSize(img): 0.0642/32
* Results *
  # test persons: 10000
  (instance-based)  accuracy:      71.8%
  (instance-based)  precition:     87.1%
  (instance-based)  recall:        78.3%
  (instance-based)  f1-score:      82.5%
  (label-based)     mean accuracy: 70.2%
  mA for each attribute: [0.8536271  0.67568934 0.6094301  0.51071817 0.8728046  0.8385074
 0.8989417  0.56558436 0.77835524 0.6497474  0.62305665 0.6803385
 0.49994949 0.86991173 0.86991173 0.7700514  0.718978   0.6398968
 0.7076544  0.5        0.6005873  0.6147441  0.8619995  0.79627764
 0.73238236 0.5       ]
`


i trained with a batch size of 256 and all the rest of the parameters were left at default values. These results are slightly off the ones from the paper but are close enough ... i guess; Do you reckon if it be possible to build a network that does both attribute recognition and person reidentification?

**Comment by KaiyangZhou** [comment]
You could try. A simple way is to build two branches on the top of the backbone, one for attribute and the other for reid

**Comment by MuhammadAsadJaved** [comment]
@vakkov  Hi there, Are you able to do two tasks (attribute recognition and person reidentification)? 
I am also trying something the same. I want to train OSNet for the person and car reidentification.  Train model on MSMT17 and VeRi datasets. Can you suggest some amendments?
